walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,211 @@
+    TimeRange meetingTime = TimeRange.fromStartEnd(0, 0, false);
+
+    blockedTimes = getBlockedTimes(events, request, considerOptional);
+
+    // if there are no other events at this point the meeting can happen whenever in the day
+    if (blockedTimes.size() == 0 && request.getOptionalAttendees().size() == 0) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    int startTime = 0;
+    int endTime = 0;
+
+    // the first time slot that should always be tested is from [0, the start time of the first event)
+    if(blockedTimes.iterator().hasNext()){
+        endTime = blockedTimes.iterator().next().start();
+    }
+ 
+    boolean isInbetween = getIsInbetween(blockedTimes, startTime);
+    
+    // if startTime and endTime are not inbetween any other blocks and longer 
+    // than request duration then it can be added to returnTimeRange
+    if (!isInbetween && endTime - startTime >= request.getDuration()) {
+      meetingTime = TimeRange.fromStartEnd(startTime, endTime, false);
+      returnTimeRange.add(meetingTime);
+    }
+
+	// for each TimeRange in blockedTimes use the end time of the TimeRange to 
+    // be new possible start time, get an endTime from next possible TimeRange, 
+    // check if these new times are inbetween blocks, if they are not add it returnTimeRange
+    for (TimeRange time : blockedTimes) {
+      startTime = time.end();
+
+      endTime = getEndTime(blockedTimes, startTime);
+
+      isInbetween = getIsInbetween(blockedTimes, startTime);
+
+      // if startTime and endTime are not inbetween any other blocks and longer 
+      // than request duration then it can be added to returnTimeRange
+      if (!isInbetween && endTime - startTime >= request.getDuration()) {
+        meetingTime = TimeRange.fromStartEnd(startTime, endTime, false);
+
+        // there are cases where the same time range can be added, this avoid that
+        boolean alreadyHaveTimeRange = getAlreadyHaveTimeRange(meetingTime, returnTimeRange);
+        if (!alreadyHaveTimeRange) {
+          returnTimeRange.add(meetingTime);
+        }
+      }
+    }
+    return returnTimeRange;
+  }
+
+ /**
+  * Gets collection of TimeRange of blocked of times given events of attendees, 
+  * the request, and if optional attendees should be considered 
+  * 
+  * @param events the events that the attendees have to visit
+  * @param request the request that is made including the attendees visiting
+  * @param considerOptional whether or not optional attendees are considered
+  * @return Collection of TimeRanges that are blocked for attendees
+  */
+  private Collection<TimeRange> getBlockedTimes(
+    Collection<Event> events,
+    MeetingRequest request,
+    boolean considerOptional
+  ) {
+    List<TimeRange> blockedTimes = new ArrayList<TimeRange>();
+    for (Event event : events) {
+      for (String eventAttendee : event.getAttendees()) {
+        // spilt into two if statements because it was unreadable as one
```

Comment:
I like that you're thinking about readability and trying to make it better with comments. But comments can often go stale, when the code changes the comment might not get updated. In this case I would add extra variables to improve readability, e.g.

```
// replace with descriptive variable names
boolean someVar = request.getAttendees().contains(eventAttendee);
boolean someOtherVar = considerOptional && request.getOptionalAttendees().contains(eventAttendee);

if (someVar || someOtherVar) {
...
```